client/app/pages/queries/hooks/useQueryExecute.js

```diff
@@ -1,79 +1,120 @@
 
-  const [isExecutionCancelling, setIsExecutionCancelling] = useState(false);
+      setExecutionState({
+        updatedAt: newQueryResult.getUpdatedAt(),
+        executionStatus: newQueryResult.getStatus(),
+        isExecuting: true,
+        cancelCallback: () => {
+          recordEvent("cancel_execute", "query", query.id);
+          setExecutionState({ isCancelling: true });
+          newQueryResult.cancelExecution();
+        },
+      });
 
-  const executeQuery = useCallback(() => {
-    recordEvent("execute", "query", query.id);
-    setQueryResult(query.getQueryResult(0));
-  }, [query]);
+      const onStatusChange = status => {
+        if (queryResultInExecution.current === newQueryResult) {
+          setExecutionState({ updatedAt: newQueryResult.getUpdatedAt(), executionStatus: status });
+        }
+      };
 
-  const executeAdhocQuery = useCallback(
-    selectedQueryText => {
-      recordEvent("execute", "query", query.id);
-      setQueryResult(query.getQueryResultByText(0, selectedQueryText));
+      newQueryResult
+        .toPromise(onStatusChange)
+        .then(queryResult => {
+          if (queryResultInExecution.current === newQueryResult) {
+            // TODO: this should probably belong in the QueryEditor page.
+            if (queryResult && queryResult.query_result.query === query.query) {
+              query.latest_query_data_id = queryResult.getId();
+              query.queryResult = queryResult;
+            }
+
+            setExecutionState({
+              queryResult,
+              loadedInitialResults: true,
+              error: null,
+              isExecuting: false,
+              isCancelling: false,
+              executionStatus: null,
+            });
+          }
+        })
+        .catch(queryResult => {
+          if (queryResultInExecution.current === newQueryResult) {
+            setExecutionState({
+              queryResult,
+              loadedInitialResults: true,
+              error: queryResult.getError(),
+              isExecuting: false,
+              isCancelling: false,
+              executionStatus: ExecutionStatus.FAILED,
+            });
+          }
+        });
     },
     [query]
   );
 
-  const cancelExecution = useCallback(() => {
-    if (queryResult) {
-      recordEvent("cancel_execute", "query", query.id);
-      queryResult.cancelExecution();
-      setIsExecutionCancelling(true);
-    }
-  }, [query.id, queryResult]);
+  const queryRef = useRef(query);
+  const executeQueryRef = useRef(executeQuery);
+  useEffect(() => {
+    queryRef.current = query;
+    executeQueryRef.current = executeQuery;
+  }, [executeQuery, query]);
 
   useEffect(() => {
-    if (!isQueryExecuting) {
-      setIsExecutionCancelling(false);
-      if (queryResult && queryResult.query_result.query === query.query) {
-        query.latest_query_data_id = queryResult.getId();
-        query.queryResult = queryResult;
-      }
+    // TODO: this belongs on the query page?
```

Comment:
I tend to prefer this logic in here mainly for 2 reasons: 1 - Reduce complexity from Pages; 2 - It's a shared piece of code between View and Source.

In terms of where this code "belongs", it feels to me it could be in both the query page or in here.

Reply:
It's not for this PR, but eventually it makes sense to use `useQueryExecute` in Widgets, where this logic isn't needed.